python/ray/tune/session.py

```diff
@@ -67,9 +75,75 @@ def init(reporter, ignore_reinit_error=True):
             "Most session commands will have no effect."
         )
 
+    # Setup hooks for generating placement group resource deadlock warnings.
+    from ray import actor, remote_function
+
+    if "TUNE_DISABLE_RESOURCE_CHECKS" not in os.environ:
+        actor._actor_launch_hook = tune_task_and_actor_launch_hook
+        remote_function._task_launch_hook = tune_task_and_actor_launch_hook
+
     _session = reporter
 
 
+# Cache of resource dicts that have been checked by the launch hook already.
+_checked_resources: Set[frozenset] = set()
+
+
+def tune_task_and_actor_launch_hook(
+    fn, resources: Dict[str, float], strategy: Optional[SchedulingStrategyT]
+):
+    """Launch hook to catch nested tasks that can't fit in the placement group.
+
+    This gives users a nice warning in case they launch a nested task in a Tune trial
+    without reserving resources in the trial placement group to fit it.
+    """
+
+    # Already checked, skip for performance reasons.
+    key = frozenset({(k, v) for k, v in resources.items() if v > 0})
+    if not key or key in _checked_resources:
+        return
+
+    # No need to check if placement group is None.
+    if (
+        not isinstance(strategy, PlacementGroupSchedulingStrategy)
+        or strategy.placement_group is None
+    ):
+        return
+
+    # Check if the resource request is targeting the current placement group.
+    cur_pg = ray.util.get_current_placement_group()
+    if not cur_pg or strategy.placement_group.id != cur_pg.id:
+        return
+
+    _checked_resources.add(key)
+
+    # Check if the request can be fulfilled by the current placement group.
+    if _valid_resource_shape(resources, cur_pg.bundle_specs[1:]):
```

Comment:
Hmmm, this may not be true. Sometimes one can specify a resourceless head for trainable.

```
def g(*a):
    @ray.remote(num_cpus=1)
    class Actor:
        def f(self):
            pass

    a = Actor.remote()
    ray.get(a.f.remote())

tune.run(
    g, resources_per_trial=tune.PlacementGroupFactory([{}, {"CPU": 1}]), verbose=0
)
```

This should be ok, but will be failed under current implementation.

PGF stuff is a bit messy right now. But it allows user to specify trainable without any resources allocated to the head. However as PG API doesn't allow empty bundle, the empty bundle is purposely removed before passing to PG.

Reply:
Yep, we only check nested tasks (ignoring the first bundle / driver).

Reply:
Sorry.. so should the snippet I pasted legit? It's currently throwing exception in current implementation.

Reply:
I see. This is a little bit messy. Why don't we just prohibit users from passing an empty resource bundle? That will cause all sorts of problems like infinite number of concurrent trials.

Reply:
Yeah agree it's a bit messy. @krfricke to provide more context as to why.
My understanding is this is really just for distributed trainable - one can opt to put all the resources on workers and not trainable itself. There won't be a problem of infinite number of trials as the minimum Tune requests is 1 CPU. 
Actually in that case can a user just specify [{"CPU": 0.001}, {"CPU": 0.999}, {"CPU": 1}]? Will that be effectively the same? This way we can get rid of this confusing semantics.

Reply:
I think we should keep the possibility of passing an empty bundle. It allows the users to easily specify that the trainable actor should request no resources itself. Forcing users to do messy fractions when they want to do something that is easily possible within core doesn't seem right to me. We only allow the first bundle to be empty, and it is always going to be used as the resources for the trainable actor. Only the first bundle, used for the trainable, can be empty. If you are worried about infinite trials, we can always just validate inside PGF if there is just the empty bundle.